python-package/lightgbm/basic.py

```diff
@@ -0,0 +1,1080 @@
+                c_str(model_file), 
+                ctypes.byref(out_num_total_model),
+                ctypes.byref(self.handle)))
+            self.__num_total_model = out_num_total_model.value
+            out_num_class = ctypes.c_int64(0)
+            _safe_call(_LIB.LGBM_BoosterGetNumClasses(
+                self.handle,
+                ctypes.byref(out_num_class)))
+            self.__num_class = out_num_class.value
+        else:
+            raise TypeError('At least need training dataset or model file to create booster instance')
+
+    def __del__(self):
+        if self.handle is not None and self.__is_manage_handle:
+            _safe_call(_LIB.LGBM_BoosterFree(self.handle))
+
+    def add_valid(self, data, name):
+        if data.predictor is not self.init_predictor:
+            raise Exception("Add validation data failed, you should use same predictor for these data")
+        _safe_call(_LIB.LGBM_BoosterAddValidData(
+            self.handle,
+            data.handle))
+        self.valid_sets.append(data)
+        self.name_valid_sets.append(name)
+        self.__num_dataset += 1
+
+    def reset_parameter(self, params, silent=False):
+        self.__need_reload_eval_info = True
+        if silent:
+            params["verbose"] = 0
+        else:
+            params["verbose"] = 1
+        params_str = dict_to_str(params)
+        _safe_call(_LIB.LGBM_BoosterResetParameter(
+            self.handle,
+            c_str(params_str)))
+
+    def update(self, train_set=None, fobj=None):
+        """
+        Update for one iteration
+        Note: for multi-class task, the score is group by class_id first, then group by row_id
+              if you want to get i-th row score in j-th class, the access way is score[j*num_data+i]
+              and you should group grad and hess in this way as well
+        Parameters
+        ----------
+        train_set : training data, None means use last training data
+        fobj : function
+            Customized objective function.
+
+        Returns
+        -------
+        is_finished, bool
+        """
+        """need reset training data"""
+        if train_set is not None and train_set is not self.train_set:
+            if train_set.predictor is not self.init_predictor:
+                raise Exception("Replace training data failed, you should use same predictor for these data")
+            self.train_set = train_set
+            _safe_call(_LIB.LGBM_BoosterResetTrainingData(
+                self.handle, 
+                self.train_set.handle))
+            self.__inner_predict_buffer[0] = None
+        is_finished = ctypes.c_int(0)
+        if fobj is None:
+            _safe_call(_LIB.LGBM_BoosterUpdateOneIter(
+                self.handle, 
+                ctypes.byref(is_finished)))
+            return is_finished.value == 1
+        else:
+            grad, hess = fobj(self.__inner_predict(0), self.train_set)
+            return self.__boost(grad, hess)
```

Comment:
Is it too strict to let user define a function with return of two 1d numpy array with dtype=float32. Maybe add some type conversions here?